### PR TITLE
bazel improvements

### PR DESCRIPTION
**Pull request by nerdrew**
- bazel paths should be relative to `bazel info output_base`
- update vim instructions for vim-lsc
- add instructions for vim-ale

### CODE REVIEW

File: src/main/java/org/javacs/InferConfig.java [hunk 1]
```diff
@@ -240,12 +257,33 @@ private static String findExecutableOnPath(String name) {
 
     private Set<Path> bazelClasspath() {
         try {
+            // Run bazel as a subprocess
+            String[] output_base_command = {"bazel", "info", "output_base"};
```
Comment (nerdrew) [inline src/main/java/org/javacs/InferConfig.java]: `output_base` seems to work, but I'm not sure if it is the correct directory. Would it be better to use `execution_base`?

Comment (nerdrew) [reply src/main/java/org/javacs/InferConfig.java]: oops. I meant `execution_root`.

**Comment by nerdrew** [comment]
I ran into some issues when testing bazel + java-language-server with our large java repo. I usually only work in a subdirectory of the workspace, hence the addition of scoping the targets to the current directory instead of the whole project.

Also, we have a bunch of external maven dependencies which aren't correctly resolved right now. I think that it works right now since most of the deps are in `bazel-out` which happens to both be a link from the workspace root to a directory of the same name in `output_base`. The external dependencies however don't work.

The external deps show up as `/external/maven/v1/https/...` for me, but there is no `external` symlink in the workspace root.

Here are the bazel symlinks in my workspace root:

```
 ls -l bazel-*
 bazel-bin -> /private/var/tmp/_bazel_blah/.../execroot/__main__/bazel-out/darwin-fastbuild/bin
bazel-genfiles -> /private/var/tmp/_bazel_blah/.../execroot/__main__/bazel-out/darwin-fastbuild/bin
bazel-java -> /private/var/tmp/_bazel_blah/.../execroot/__main__
bazel-out -> /private/var/tmp/_bazel_blah/.../execroot/__main__/bazel-out
bazel-testlogs -> /private/var/tmp/_bazel_blah/.../execroot/__main__/bazel-out/darwin-fastbuild/testlogs`
```

**Comment by nerdrew** [comment]
The warnings I have left all look like:

```
14:49:03.139    WARNING org.javacs.Classes jdkTopLevelClasses   Failed indexing module java.activation(/modules/java.activation)
```

File: src/main/java/org/javacs/InferConfig.java [hunk 2]
```diff
@@ -22,20 +22,33 @@
     private final Path mavenHome;
     /** Location of the gradle cache, usually ~/.gradle */
     private final Path gradleHome;
+    /** VSCode's current working directory */
+    private final String target;
 
-    InferConfig(Path workspaceRoot, Collection<String> externalDependencies, Path mavenHome, Path gradleHome) {
+    InferConfig(
```
Comment (georgewfraser) [inline src/main/java/org/javacs/InferConfig.java]: Why is workspaceroot different than cwd? My understanding of how VSCode works is they should always be the same.

Comment (nerdrew) [reply src/main/java/org/javacs/InferConfig.java]: I've been using this with vim + ale. We have a large monorepo, so my workflow often involves working in a subdirectory of the workspace root. I haven't tested with VSCode. I assume you'd be able to use VSCode in a similar manner though.

Comment (georgewfraser) [reply src/main/java/org/javacs/InferConfig.java]: As far as I can tell, the cwd that the language server process runs in isn't part of the LSP, so we shouldn't rely on it for anything. When you're working in a subdirectory of a large monorepo, workspaceRoot should be the subdirectory, not the parent directory.

Comment (nerdrew) [reply src/main/java/org/javacs/InferConfig.java]: Does the code currently walk up from the workspaceRoot looking for the bazel WORKSPACE file? I think if we do that, then leaving the workspaceRoot == cwd will still work for me and allow me to add the relative path from the bazel WORKSPACE -> workspaceRoot in the bazel commands. Does that make sense?

Comment (georgewfraser) [reply src/main/java/org/javacs/InferConfig.java]: I think that's what it does. If it doesn't, it should.

**Comment by georgewfraser** [comment]
@nerdrew If you refactor this to eliminate use of `cwd`, I'd be happy to merge.

**Comment by nerdrew** [comment]
Removing the cwd code now. I'm going to verify that it walks up the directories looking for the bazel `WORKSPACE` file.

File: src/main/java/org/javacs/InferConfig.java (hunk 2 above)
Comment (nerdrew) [reply src/main/java/org/javacs/InferConfig.java]: It doesn't seem to recursively search parent directories for the bazel `WORKSPACE` file. It looks like it checks the `workspaceRoot` only. At least, that's all it appeared to do for me.

**Comment by nerdrew** [comment]
Removed all the cwd code. I didn't fix finding the bazel `WORKSPACE` file yet. I think that can merge separately though.

**Comment by nerdrew** [comment]
Ok. Added a commit to find the bazel `WORKSPACE` file in the parent tree.

**Comment by georgewfraser** [comment]
Sorry I haven't looked at this in a while. Is this ready-to-merge @nerdrew ?

**Comment by georgewfraser** [comment]
This has gotten somewhat out-of-sync with master, so I took some of the changes and put them in https://github.com/georgewfraser/java-language-server/commit/337fa077c8670f21ca9808f98e7da012131cd43e. Thank you for contributing! If I've missed something important, please make a new PR.

**Comment by nerdrew** [comment]
Thanks for picking up my changes. Sorry I didn't respond before :/

**Comment by nerdrew** [comment]
There actually were a couple changes that I was hoping to get included. I'll open a new PR.

**Pull request closed without merging**